extensions/amp-fx-parallax/0.1/amp-fx-parallax.js

```diff
@@ -0,0 +1,223 @@
+  }
+
+  /**
+   * Install handlers after the document has loaded.
+   */
+  start_() {
+    const win = this.win_;
+    onDocumentReady(win.document, () => {
+      installParallaxHandlers(win);
+    });
+  }
+
+  /**
+   * Get the previous scroll value.
+   * @return {number}
+   */
+  getPreviousScroll() {
+    return this.previousScroll_;
+  }
+
+  /**
+   * Set the previous scroll value.
+   * @param {number} scroll
+   */
+  setPreviousScroll(scroll) {
+    this.previousScroll_ = scroll;
+  }
+
+  /**
+   * Add listeners to parallax scroll events.
+   * @param {!function()} cb
+   */
+  addScrollListener(cb) {
+    this.parallaxObservable_.add(cb);
+  }
+
+  /**
+   * Remove listeners from parallax scroll events.
+   * @param {!function()} cb
+   */
+  removeScrollListener(cb) {
+    this.parallaxObservable_.remove(cb);
+  }
+
+  /**
+   * Alert listeners that a scroll has occurred.
+   * @param {number} scrollTop
+   */
+  fire(scrollTop) {
+    this.parallaxObservable_.fire(scrollTop);
+  }
+}
+
+/**
+ * Encapsulates and tracks an element's linear parallax effect.
+ */
+export class ParallaxElement {
+  /**
+   * @param {!Element} element The element to give a parallax effect.
+   * @param {!function(number):string} transform Computes the transform from the position.
+   */
+  constructor(element, transform) {
+    const factor = element.getAttribute(ATTR);
+
+    /** @private @const {!Element} */
+    this.element_ = element;
+
+    /** @private @const {!function(number):string} */
+    this.transform_ = transform;
+
+    /** @private @const {number} */
+    this.factor_ = (factor ? parseFloat(factor) : 0.5) - 1;
+
+    /** @private {number} */
+    this.offset_ = 0;
+  }
+
+  /**
+   * Apply the parallax effect to the offset given how much the page
+   * has moved since the last frame.
+   * @param {number} delta The movement of the base layer e.g. the page.
+   */
+  update(delta) {
+    this.offset_ += delta * this.factor_;
+    setStyles(this.element_, {transform: this.transform_(this.offset_)});
+  }
+
+  /**
+   * True if the element is in the viewport.
+   * @param {!../../../src/service/viewport-impl.Viewport} viewport
+   * @return {boolean}
+   */
+  shouldUpdate(viewport) {
+    const viewportRect = viewport.getRect();
+    const elementRect = viewport.getLayoutRect(this.element_);
+    elementRect.top -= viewportRect.top;
+    elementRect.bottom = elementRect.top + elementRect.height;
+    return this.isRectInView_(elementRect, viewportRect.height);
+  }
+
+  /**
+   * Check if a rectange is within the viewport.
+   * @param {!../../../src/layout-rect.LayoutRectDef} rect
+   * @param {number} viewportHeight
+   * @private
+   */
+  isRectInView_(rect, viewportHeight) {
+    return rect.bottom >= 0 && rect.top <= viewportHeight;
+  }
+}
+
+/**
+ * Constructs and installs scroll handlers on all [amp-fx-parallax] elements
+ * in the document.
+ * @param {!Window} global
+ */
+function installParallaxHandlers(global) {
```

Comment:
let's move this, `parallaxMutate` and `transform` to be private methods on `ParallaxService` class and make all other methods on `ParallaxService` also private.